File: seeder lama/seed/shop_seeder.py
from typing import List, Dict, Any
from sqlalchemy.orm import Session

try:
    from src.models.shop_items import ShopItem, ShopItemType
    from src.database.seeder import BaseSeeder
except ImportError:
    from ...models.shop_items import ShopItem, ShopItemType
    from ..seeder import BaseSeeder

class ShopSeeder(BaseSeeder):
    """Seeder untuk Shop Items"""

    def __init__(self):
        super().__init__()

    def run(self):
        try:
            print("🌱 Starting shop item seeding...")
            self.seed_shop_items()
            print("✅ Shop item seeding finished successfully!")
        except Exception as e:
            self.db.rollback()
            raise Exception(f"Shop item seeding failed: {e}")

    def seed_shop_items(self):
        print("🛒 Seeding Shop Items...")
        items_data = [
            {
                "name": "Streak Freeze",
                "description": "Melindungi streak harian agar tidak hilang.",
                "item_type": ShopItemType.STREAK_FREEZE,
                "xp_cost": 100,
                "icon": "streak_freeze.png"
            },
            {
                "name": "Badge Emas",
                "description": "Badge spesial untuk pencapaian emas.",
                "item_type": ShopItemType.BADGE,
                "xp_cost": 250,
                "icon": "badge_emas.png"
            },
            {
                "name": "XP Boost",
                "description": "Dapatkan XP tambahan selama 1 jam.",
                "item_type": ShopItemType.BOOST,
                "xp_cost": 150,
                "icon": "xp_boost.png"
            },
        ]

        created_count = 0
        for data in items_data:
            existing = self.db.query(ShopItem).filter(ShopItem.name == data["name"]).first()
            if not existing:
                item = ShopItem(**data)
                self.db.add(item)
                created_count += 1
                print(f"  ✅ Created shop item: {data['name']}")
            else:
                print(f"  ⚠️ Shop item already exists: {data['name']}")

        self.db.commit()
        print(f"✅ Shop item seeding completed. Created {created_count} items.")

if __name__ == "__main__":
    ShopSeeder().run()

File: cmd/seed/seeder/shop_seeder.go
package seeder

import (
    "database/sql"
    "fmt"
    "github.com/jmoiron/sqlx"
    "REFACTORING_MAUNA/internal/domain/entities"
)

type ShopSeeder struct {
    db *sqlx.DB
}

func NewShopSeeder(db *sqlx.DB) *ShopSeeder {
    return &ShopSeeder{db: db}
}

func (s *ShopSeeder) Name() string {
    return "ShopSeeder"
}

type shopItemData struct {
    Name        string
    Description string
    ItemType    entities.ShopItemType
    XpCost      int
    Icon        string
}

func (s *ShopSeeder) Run() error {
    PrintInfo("Starting shop item seeding...")

    itemsData := []shopItemData{
        {
            Name:        "Streak Freeze",
            Description: "Melindungi streak harian agar tidak hilang.",
            ItemType:    entities.TypeStreakFreeze,
            XpCost:      100,
            Icon:        "streak_freeze.png",
        },
        {
            Name:        "Badge Emas",
            Description: "Badge spesial untuk pencapaian emas.",
            ItemType:    entities.TypeBadge,
            XpCost:      250,
            Icon:        "badge_emas.png",
        },
        {
            Name:        "XP Boost",
            Description: "Dapatkan XP tambahan selama 1 jam.",
            ItemType:    entities.TypeBoost,
            XpCost:      150,
            Icon:        "xp_boost.png",
        },
    }

    // Start transaction
    tx, err := s.db.Beginx()
    if err != nil {
        PrintError(fmt.Sprintf("Failed to start transaction: %v", err))
        return err
    }
    defer func() {
        if err != nil {
            tx.Rollback()
        }
    }()

    PrintInfo("Seeding Shop Items...")
    createdCount := 0

    for _, item := range itemsData {
        // Check if shop item already exists
        var existingID sql.NullInt64
        err := tx.QueryRow(
            "SELECT id FROM shop_items WHERE name = $1",
            item.Name,
        ).Scan(&existingID)

        if err == nil && existingID.Valid {
            // Item already exists
            PrintWarning(fmt.Sprintf("Shop item already exists: %s", item.Name))
            continue
        }

        if err != nil && err != sql.ErrNoRows {
            // Unexpected error
            PrintError(fmt.Sprintf("Database error: %v", err))
            return err
        }

        // Insert new shop item
        _, err = tx.Exec(
            "INSERT INTO shop_items (name, description, item_type, xp_cost, icon) VALUES ($1, $2, $3, $4, $5)",
            item.Name,
            item.Description,
            item.ItemType,
            item.XpCost,
            item.Icon,
        )

        if err != nil {
            PrintError(fmt.Sprintf("Failed to create shop item: %v", err))
            return err
        }

        PrintSuccess(fmt.Sprintf("Created shop item: %s", item.Name))
        createdCount++
    }

    // Commit transaction
    if err := tx.Commit(); err != nil {
        PrintError(fmt.Sprintf("Failed to commit transaction: %v", err))
        return err
    }

    PrintSuccess(fmt.Sprintf("Shop item seeding completed. Created %d items.", createdCount))
    return nil
}